great_expectations/experimental/datasources/interfaces.py

```diff
@@ -63,7 +74,12 @@ def get_batch_request(self, options: Optional[BatchRequestOptions]) -> BatchRequ
         raise NotImplementedError
 
 
-class Datasource(ExperimentalBaseModel, metaclass=MetaDatasource):
+DataAssetType = TypeVar("DataAssetType", bound=DataAsset)
+
+
+class Datasource(
+    ExperimentalBaseModel, Generic[DataAssetType], metaclass=MetaDatasource
+):
```

Comment:
So now the type-system knows that a `PostgresDatasource` contains `TableAsset`s and not a general/non-specific `DataAsset` type?
🙌